### PR TITLE
create Relation.Binary.Construct.Closure.Reflexive.Properties

### PR DESCRIPTION
**Pull request by strake**

### CODE REVIEW

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    irrel∧irrefl⟶irrel _ ~-irrefl [ x∼y ] refl = ⊥-elim (~-irrefl refl x∼y)
+    irrel∧irrefl⟶irrel _ ~-irrefl refl [ x∼y ] = ⊥-elim (~-irrefl refl x∼y)
+    irrel∧irrefl⟶irrel _ _ refl refl = refl
+
+    asym⟶antisym : ∀ {ℓ'} {_≈_ : Rel _ ℓ'} → Reflexive _≈_ → Asymmetric _~_ → Antisymmetric _≈_ (Refl _~_)
+    asym⟶antisym ≈-refl ~-asym [ x∼y ] [ x∼y′ ] = ⊥-elim (~-asym x∼y x∼y′)
+    asym⟶antisym ≈-refl _ _ refl = ≈-refl
+    asym⟶antisym ≈-refl _ refl _ = ≈-refl
+
+    partialOrder : IsStrictPartialOrder _≡_ _~_ → IsPartialOrder _≡_ (Refl _~_)
```
Comment (MatthewDaggitt) [inline src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda]: These should be called `isPartialOrder`, `isDecPartialOrder` etc.

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    partialOrder ~-IsStrictPartialOrder = record
+        { isPreorder = record
```
Comment (MatthewDaggitt) [inline src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda]: Can you split `isPreorder` out please?

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    decPartialOrder ~-IsDecStrictPartialOrder = record
+      { isPartialOrder = partialOrder isStrictPartialOrder; _≟_ = _≟_; _≤?_ = dec _≟_ _<?_ }
+      where open IsDecStrictPartialOrder ~-IsDecStrictPartialOrder
+
+    totalOrder : IsStrictTotalOrder _≡_ _~_ → IsTotalOrder _≡_ (Refl _~_)
+    totalOrder ~-IsStrictTotalOrder = record
+      { isPartialOrder = partialOrder isStrictPartialOrder; total = total }
+      where open IsStrictTotalOrder ~-IsStrictTotalOrder
+
+            total : Total (Refl _~_)
```
Comment (MatthewDaggitt) [inline src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda]: This feels like it should go in its own top-level proof

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    asym⟶antisym ≈-refl _ refl _ = ≈-refl
+
+    partialOrder : IsStrictPartialOrder _≡_ _~_ → IsPartialOrder _≡_ (Refl _~_)
+    partialOrder ~-IsStrictPartialOrder = record
+        { isPreorder = record
+            { isEquivalence = isEquivalence
+            ; reflexive = λ { refl → refl }
+            ; trans = trans (IsStrictPartialOrder.trans ~-IsStrictPartialOrder) }
+        ; antisym = asym⟶antisym {_} {_≡_} Eq.refl asym }
+      where open IsStrictPartialOrder ~-IsStrictPartialOrder hiding (trans)
```
Comment (MatthewDaggitt) [inline src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda]: The neatest way to handle this situation we've found is to write `where module O = IsStrictPartialOrder ~-IsStrictPartialOrder` and then write `O.trans`, `O.asym` etc...

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+            total : Total (Refl _~_)
+            total x y with compare x y
+            total _ _ | tri< a _ _ = inj₁ [ a ]
+            total _ _ | tri≈ _ refl _ = inj₁ refl
+            total _ _ | tri> _ _ c = inj₂ [ c ]
+
+    decTotalOrder : IsStrictTotalOrder _≡_ _~_ → IsDecTotalOrder _≡_ (Refl _~_)
+    decTotalOrder ~-IsStrictTotalOrder = record
+      { isTotalOrder = totalOrder ~-IsStrictTotalOrder; _≟_ = _≟_; _≤?_ = dec _≟_ _<?_ }
```
Comment (MatthewDaggitt) [inline src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda]: Please could you lay the record out as per the style guide? New entry in it recently!

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    asym⟶antisym ≈-refl _ _ refl = ≈-refl
+    asym⟶antisym ≈-refl _ refl _ = ≈-refl
+
+    partialOrder : IsStrictPartialOrder _≡_ _~_ → IsPartialOrder _≡_ (Refl _~_)
+    partialOrder ~-IsStrictPartialOrder = record
+        { isPreorder = record
+            { isEquivalence = isEquivalence
+            ; reflexive = λ { refl → refl }
+            ; trans = trans (IsStrictPartialOrder.trans ~-IsStrictPartialOrder) }
+        ; antisym = asym⟶antisym {_} {_≡_} Eq.refl asym }
```
Comment (MatthewDaggitt) [inline src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda]: Looks like the equality argument in `asym⟶antisym` needs to be explicit?

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda (hunk 3 above)
Comment (strake) [reply src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda]: I think `total` requires `IsStrictTotalOrder` anyhow, so it's not really a win — one can rather merely use `IsTotalOrder.total (isTotalOrder ?)`.

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda (hunk 1 above)
Comment (strake) [reply src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda]: Done

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda (hunk 2 above)
Comment (strake) [reply src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda]: Done

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda (hunk 5 above)
Comment (strake) [reply src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda]: Done

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda (hunk 4 above)
Comment (strake) [reply src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda]: Now unnecessary, i believe

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda (hunk 6 above)
Comment (strake) [reply src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda]: Where? It checks out fine for me.

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties.agda [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+{-# OPTIONS --safe --without-K #-}
+
+module Relation.Binary.Construct.Closure.Reflexive.Properties where
+
+open import Data.Product as Prod
+open import Function
+open import Relation.Binary
+open import Relation.Binary.Construct.Closure.Reflexive
+import Relation.Binary.PropositionalEquality as PropEq
+open PropEq using (_≡_; refl)
```
Comment (MatthewDaggitt) [inline src/Relation/Binary/Construct/Closure/Reflexive/Properties.agda]: No need to split the `import` and `open` statements here

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties.agda [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+open import Relation.Nullary
+
+module _ {a b ℓ} {A : Set a} {B : Set b} {_~_ : Rel A ℓ} where
```
Comment (MatthewDaggitt) [inline src/Relation/Binary/Construct/Closure/Reflexive/Properties.agda]: Everything should be indented using only two spaces. You could also consider using the new `variable` keyword for `a,b,l,A,B`.

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties.agda [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    =[]⇒ : ∀ {p q} {P : Rel A p} {Q : Rel B q} {f} → P =[ f ]⇒ Q → Refl P =[ f ]⇒ Refl Q
+    =[]⇒ x [ x∼y ] = [ x x∼y ]
+    =[]⇒ x refl = refl
```
Comment (MatthewDaggitt) [inline src/Relation/Binary/Construct/Closure/Reflexive/Properties.agda]: Please could you try and align the `=` sign where possible, makes it much easier to read.

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties.agda [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+    respˡ : ∀ {p} {P : REL A B p} → P Respectsˡ _~_ → P Respectsˡ (Refl _~_)
+    respˡ p-respˡ-~ [ x∼y ] = p-respˡ-~ x∼y
+    respˡ _ refl = id
+
+    respʳ : ∀ {p} {P : REL B A p} → P Respectsʳ _~_ → P Respectsʳ (Refl _~_)
+    respʳ = respˡ
+
+module _ {a ℓ} {A : Set a} {_~_ : Rel A ℓ} where
+    dec : Decidable {_} {A} _≡_ → Decidable _~_ → Decidable (Refl _~_)
```
Comment (MatthewDaggitt) [inline src/Relation/Binary/Construct/Closure/Reflexive/Properties.agda]: Writing this as `Decidable {A = A} _≡_` is a little neater and clearer what's going on.

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties.agda [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+    dec ≡-dec ~-dec a b with ≡-dec a b | ~-dec a b
+    dec ≡-dec ~-dec _ _ | _ | yes q = yes [ q ]
```
Comment (MatthewDaggitt) [inline src/Relation/Binary/Construct/Closure/Reflexive/Properties.agda]: Please replace duplicated code after `with` statements with `...` and align the remaining `|`s where possible.

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda (hunk 3 above)
Comment (MatthewDaggitt) [reply src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda]: Looks to me like it only requires `Trichotomous`? You can use `IsTotalOrder.total (isTotalOrder ?)` but it's considerably longer than `total`.

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+    irrel∧irrefl⟶irrel _ ~-irrefl refl [ x∼y ] = ⊥-elim (~-irrefl refl x∼y)
+    irrel∧irrefl⟶irrel _ _ refl refl = refl
+
+    asym⟶antisym : ∀ {ℓ'} {_≈_ : Rel _ ℓ'} → Reflexive _≈_ → Asymmetric _~_ → Antisymmetric _≈_ (Refl _~_)
+    asym⟶antisym ≈-refl ~-asym [ x∼y ] [ x∼y′ ] = ⊥-elim (~-asym x∼y x∼y′)
+    asym⟶antisym ≈-refl _ _ refl = ≈-refl
+    asym⟶antisym ≈-refl _ refl _ = ≈-refl
+
+    isPreorder : Transitive _~_ → IsPreorder _≡_ (Refl _~_)
+    isPreorder ~-trans = record
```
Comment (MatthewDaggitt) [inline src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda]: Please can you align the `=` signs in the record declarations? Same thing with the proof arguments above, e.g. in `asym⟶antisym`

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+open import Data.Empty.Irrelevant using (⊥-elim)
+open import Data.Sum as Sum
+open import Relation.Binary
+open import Relation.Binary.Construct.Closure.Reflexive
+open import Relation.Binary.Construct.Closure.Reflexive.Properties public
+import Relation.Binary.PropositionalEquality as PropEq
+open PropEq using (_≡_; refl)
+
+module _ {a ℓ} {A : Set a} {_~_ : Rel A ℓ} where
+    irrel∧irrefl⟶irrel : Irrelevant _~_ → Irreflexive _≡_ _~_ → Irrelevant (Refl _~_)
```
Comment (MatthewDaggitt) [inline src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda]: Looking at other modules in `Relation.Binary.Construct` it looks like we don't typically put the preconditions in the type, and simply write `irrel` and `antisym` etc.

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda (hunk 6 above)
Comment (MatthewDaggitt) [reply src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda]: Sorry, should have been clearer! If Agda doesn't report unsolved metas when `{_} {_≡_}` is removed then they should be removed. If it does report unsolved metas when they're removed then they need to be provided explicitly, as the library encourages implicits only when they can be inferred.

**Comment by MatthewDaggitt** [comment]
Thanks for the changes, got a few more comments above!

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties.agda (hunk 7 above)
Comment (strake) [reply src/Relation/Binary/Construct/Closure/Reflexive/Properties.agda]: Unsplit

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties.agda (hunk 8 above)
Comment (strake) [reply src/Relation/Binary/Construct/Closure/Reflexive/Properties.agda]: > Everything should be indented using only two spaces.

Done

> You could also consider using the new `variable` keyword for `a,b,l,A,B`.

I couldn't figure out how to do so — if i merely use the obvious `variable` declaration, type-checking fails. Unless you feel strongly i'll just leave it as is.

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties.agda (hunk 9 above)
Comment (strake) [reply src/Relation/Binary/Construct/Closure/Reflexive/Properties.agda]: Done

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties.agda (hunk 10 above)
Comment (strake) [reply src/Relation/Binary/Construct/Closure/Reflexive/Properties.agda]: Done

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties.agda (hunk 11 above)
Comment (strake) [reply src/Relation/Binary/Construct/Closure/Reflexive/Properties.agda]: Done where possible — seems impossible with `decidable` (formerly called `tri⟶dec∘Refl`)

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda (hunk 12 above)
Comment (strake) [reply src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda]: Ah, indeed, `Total` only requires `Trichotomous`. Done and done

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda (hunk 13 above)
Comment (strake) [reply src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda]: Done — was following `Relation.Binary.Consequences`

**Comment by strake** [comment]
@MatthewDaggitt Thanks for review, PTAL

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda (hunk 6 above)
Comment (strake) [reply src/Relation/Binary/Construct/Closure/Reflexive/Properties/WithK.agda]: Done

File: src/Relation/Binary/Construct/Closure/Reflexive/Properties.agda (hunk 8 above)
Comment (MatthewDaggitt) [reply src/Relation/Binary/Construct/Closure/Reflexive/Properties.agda]: No worries, that's fine.

**Comment by MatthewDaggitt** [comment]
Thanks for the changes. I've had a fiddle around with this and had a couple of thoughts:
- As far as I can see none of the code uses axiom K so you should be able to combine the two modules (don't forget to remove the CHANGELOG entry).
- It's kind of sad that the reflexive closure isn't parameterised by an equality relation and therefore we're stuck with propositional equality. That might be something to change in the future...

**Comment by strake** [comment]
It seems the K axiom is needed to type-check `irrel` and `antisym`.

The reflexive-transitive closure isn't parametrized by an equality relation either; if we parametrize one we ought to parametrize both. But i think we can leave it as future work.

**Comment by MatthewDaggitt** [comment]
> It seems the K axiom is needed to type-check irrel and antisym.

It isn't needed for `antisym`, so I've transferred most of the properties across. Only irrelevance is left in the `WithK` module now. Looks good, I'll merge it in, in a couple of a days.

**Comment by strake** [comment]
@MatthewDaggitt

I see, thanks. My usual modus operandi is to disable the K axiom, see what breaks, and move it into a separate module 😄

I remain unsure why Agda can't infer the `_≈_` argument of `antisym`, however.

**Comment by MatthewDaggitt** [comment]
> I remain unsure why Agda can't infer the _≈_ argument of antisym, however.
 
I'm not sure why, but `Reflexive` is never sufficient for Agda to infer the relation. 

Thanks for the PR, merging in!

**Pull request merged**